tools/fork-network/src/cli.rs

```diff
@@ -1,190 +1,534 @@
+    fn mutate_state(
+        &self,
+        fork_head_block: &Block,
+        all_shard_uids: &[ShardUId],
+        home_dir: &Path,
+        store: Store,
+        near_config: &NearConfig,
+        epoch_manager: Arc<EpochManagerHandle>,
+    ) -> anyhow::Result<(Vec<StateRoot>, Vec<AccountInfo>)> {
+        let runtime = NightshadeRuntime::from_config(
+            home_dir,
+            store.clone(),
+            &near_config,
+            epoch_manager.clone(),
+        );
         let prev_state_roots = fork_head_block
             .chunks()
             .iter()
             .map(|chunk| chunk.prev_state_root())
             .collect::<Vec<_>>();
 
-        // TODO(robin-near): Make this part customizable via some kind of config file.
-        let self_validator = near_config.validator_signer.as_ref().unwrap();
-        let self_account = AccountInfo {
-            account_id: self_validator.validator_id().clone(),
-            amount: 50_000 * NEAR_BASE,
-            public_key: self_validator.public_key(),
-        };
-
-        let mut storage_mutator = StorageMutator::new(
-            epoch_manager.clone(),
-            &*runtime,
-            fork_head_block.header().epoch_id(),
-            *fork_head_block.header().prev_hash(),
-            &prev_state_roots,
-        )?;
+        let epoch_id = fork_head_block.header().epoch_id().clone();
+        let prev_hash = *fork_head_block.header().prev_hash();
+        let runtime1 = runtime.clone();
+        let make_storage_mutator: MakeSingleShardStorageMutatorFn =
+            Arc::new(move |shard_id, prev_state_root| {
+                SingleShardStorageMutator::new(
+                    shard_id,
+                    &runtime1.clone(),
+                    prev_hash,
+                    prev_state_root,
+                )
+            });
 
         let runtime_config_store = RuntimeConfigStore::new(None);
         let runtime_config = runtime_config_store.get_config(PROTOCOL_VERSION);
-        let storage_bytes = runtime_config.fees.storage_usage_config.num_bytes_account;
-        let liquid_balance = 100_000_000 * NEAR_BASE;
-        storage_mutator.set_account(
-            self_validator.validator_id().clone(),
-            Account::new(liquid_balance, self_account.amount, CryptoHash::default(), storage_bytes),
-        )?;
-        storage_mutator.set_access_key(
-            self_account.account_id.clone(),
-            self_account.public_key.clone(),
-            AccessKey::full_access(),
+
+        let tmp_state_roots = self.prepare_state(
+            &all_shard_uids,
+            store,
+            &prev_state_roots,
+            make_storage_mutator.clone(),
         )?;
+        let storage_mutator =
+            StorageMutator::new(epoch_manager, &runtime, epoch_id, prev_hash, tmp_state_roots)?;
+        let (new_state_roots, new_validator_accounts) =
+            self.add_validator_accounts(runtime_config, home_dir, storage_mutator)?;
```

Comment:
Can we separate doing mirror setup and adding validators by some command line argument? I think for new betanet setup we don't ever need to do both at the same time. At least we should have an option to only add validators fast.

Reply:
Done, 4 sub-commands available now.